server/src/main/java/org/elasticsearch/cluster/metadata/SingleNodeShutdownMetadata.java

```diff
@@ -0,0 +1,268 @@
+        this.reason = in.readString();
+        this.status = in.readEnum(Status.class);
+        this.startedAtDate = in.readVLong();
+        this.shardMigrationStatus = new ComponentShutdownStatus(in);
+    }
+
+    /**
+     * @return The ID of the node this {@link SingleNodeShutdownMetadata} concerns.
+     */
+    public String getNodeId() {
+        return nodeId;
+    }
+
+    /**
+     * @return The type of shutdown this is (shutdown vs. permanent).
+     */
+    public String getType() {
+        return type;
+    }
+
+    /**
+     * @return The user-supplied reason this node is shutting down.
+     */
+    public String getReason() {
+        return reason;
+    }
+
+    /**
+     * @return The status of this node's shutdown.
+     */
+    public Status isStatus() {
+        return status;
+    }
+
+    /**
+     * @return The timestamp that this shutdown procedure was started.
+     */
+    public long getStartedAtDate() {
+        return startedAtDate;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(nodeId);
+        out.writeString(type);
+        out.writeString(reason);
+        out.writeEnum(status);
+        out.writeVLong(startedAtDate);
+        shardMigrationStatus.writeTo(out);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        {
+            builder.field(NODE_ID_FIELD.getPreferredName(), nodeId);
+            builder.field(TYPE_FIELD.getPreferredName(), type);
+            builder.field(REASON_FIELD.getPreferredName(), reason);
+            builder.field(STATUS_FIELD.getPreferredName(), status);
+            builder.timeField(STARTED_AT_MILLIS_FIELD.getPreferredName(), STARTED_AT_READABLE_FIELD, startedAtDate);
+            builder.field(SHARD_MIGRATION_FIELD.getPreferredName(), shardMigrationStatus);
+        }
+        builder.endObject();
+
+        return builder;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if ((o instanceof SingleNodeShutdownMetadata) == false) return false;
+        SingleNodeShutdownMetadata that = (SingleNodeShutdownMetadata) o;
+        return isStatus() == that.isStatus()
+            && getStartedAtDate() == that.getStartedAtDate()
+            && getNodeId().equals(that.getNodeId())
+            && getType().equals(that.getType())
+            && getReason().equals(that.getReason());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getNodeId(), getType(), getReason(), isStatus(), getStartedAtDate());
+    }
+
+    public enum Type {
+        REMOVE,
+        RESTART
+    }
+
+    public enum Status {
+        NOT_STARTED,
+        IN_PROGRESS,
+        STALLED,
+        COMPLETE
+    }
+
+    public static class ComponentShutdownStatus extends AbstractDiffable<ComponentShutdownStatus> implements ToXContentFragment {
```

Comment:
I actually think this would be good as a separate file rather than an inner class. Makes it much easier to test/extend in the future also.

Reply:
++, I kind of figured this would get broken out, but didn't actually get around to doing so.